Tensor.random_ check that from and to are in tensor dtype bounds

pbelevich opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#34033 Tensor.random_ check that from and to are in tensor dtype bounds**

Differential Revision: [D20182414](https://our.internmc.facebook.com/intern/diff/D20182414)

pbelevich committed:
Tensor.random_ check that from and to are in tensor dtype bounds

[ghstack-poisoned]

pbelevich committed:
Update on "Tensor.random_ check that from and to are in tensor dtype bounds"


Differential Revision: [D20182414](https://our.internmc.facebook.com/intern/diff/D20182414)

[ghstack-poisoned]

pbelevich committed:
Update on "Tensor.random_ check that from and to are in tensor dtype bounds"


Differential Revision: [D20182414](https://our.internmc.facebook.com/intern/diff/D20182414)

[ghstack-poisoned]

pbelevich committed:
Update on "Tensor.random_ check that from and to are in tensor dtype bounds"


Differential Revision: [D20182414](https://our.internmc.facebook.com/intern/diff/D20182414)

[ghstack-poisoned]

anjali411 commented on aten/src/ATen/native/DistributionTemplates.h:
```diff
@@ -64,6 +64,40 @@ at::Tensor& random_impl(at::Tensor& self, at::Generator* generator) {
+  if (isFloatingType(scalar_type)) {
+    AT_DISPATCH_FLOATING_TYPES_AND2(at::ScalarType::Half, at::ScalarType::BFloat16, scalar_type, "check_random_fp_bounds", [&] {
+      const auto min = static_cast<double>(std::numeric_limits<scalar_t>::lowest());
+      const auto max = static_cast<double>(std::numeric_limits<scalar_t>::max());
```
I think we should use ```min = - std::numeric_limits<scalar_t>::infinity(),  max = std::numeric_limits<scalar_t>::infinity()``` because floating point tensors can be filled with +/- inf, so comparing them with numeric_limits::min/max will produce unexpected results

pbelevich replied:
1. `std::numeric_limits<scalar_t>::infinity()` will be casted to `std::numeric_limits<double>::infinity()` and `from < min || from > max` will become just `false` for any type and any value `from`.
2. random accepts only int64_t min and max, why do we need comparison with infinities?

anjali411 commented on aten/src/ATen/native/DistributionTemplates.h:
```diff
@@ -64,6 +64,40 @@ at::Tensor& random_impl(at::Tensor& self, at::Generator* generator) {
+    AT_DISPATCH_INTEGRAL_TYPES_AND(at::ScalarType::Bool, scalar_type, "check_random_integral_bounds", [&]() {
+      const auto min = static_cast<int64_t>(std::numeric_limits<scalar_t>::lowest());
+      const auto max = static_cast<int64_t>(std::numeric_limits<scalar_t>::max());
+      if (from < min || from > max) {
```
you could alternatively move this check to the macro and redefine it as
```#define OUT_OF_BOUNDS_WARNING_CHECK(name, min, max) { \
          if(name < min || name > max) \
               TORCH_WARN(#name, " ....") \
      }
```

anjali411 replied:
that way this call will be reduced to ```OUT_OF_BOUNDS_WARNING_CHECK(from, min, max);```

anjali411 commented on aten/src/ATen/native/DistributionTemplates.h:
```diff
@@ -64,6 +64,40 @@ at::Tensor& random_impl(at::Tensor& self, at::Generator* generator) {
+      if (from < min || from > max) {
+        show_out_of_bounds_warning("from");
+      }
+      if (to_inc < min || to_inc > max) {
```
same as above ```OUT_OF_BOUNDS_WARNING_CHECK(to_inc, min, max)```

anjali411 commented on aten/src/ATen/native/DistributionTemplates.h:
```diff
@@ -64,6 +64,40 @@ at::Tensor& random_impl(at::Tensor& self, at::Generator* generator) {
+#define show_out_of_bounds_warning(name) \
+  TORCH_WARN(name, " is out of bounds for ", dtype, ". This warning will become an error in version 1.6 release, please fix the code in advance")
+
+static void check_from_to(int64_t from, int64_t to_inc, caffe2::TypeMeta dtype) {
```
nit: maybe rename it as check_from_to_in_range

pbelevich replied:
renamed

anjali411 commented on aten/src/ATen/native/DistributionTemplates.h:
```diff
@@ -80,6 +114,7 @@ at::Tensor& random_from_to_impl(at::Tensor& self, int64_t from, c10::optional<in
         TORCH_CHECK(from < to, "random_ expects 'from' casted to dtype to be less than 'to' casted to dtype, but got from=", from, " >= to=", to);
       });
     }
+    check_from_to(from, to - 1, self.dtype());
```
if you like, you can add a comment here as to why the upper limit is to-1

pbelevich commented on aten/src/ATen/native/DistributionTemplates.h:
```diff
@@ -64,6 +64,40 @@ at::Tensor& random_impl(at::Tensor& self, at::Generator* generator) {
   return self;
 }
 
+#define show_out_of_bounds_warning(name) \
```
changed

pbelevich committed:
Update on "Tensor.random_ check that from and to are in tensor dtype bounds"


Differential Revision: [D20182414](https://our.internmc.facebook.com/intern/diff/D20182414)

[ghstack-poisoned]

anjali411 reviewed (approved):
LGTM :)
